File: cpp/leetcode/029.cpp
class Solution {
public:
    int divide(int dividend, int divisor) {
        long long res = 0;
        int sign = (dividend < 0) ^ (divisor < 0) ? -1 : 1;
        long long dvd = labs(dividend), dvs = labs(divisor);
        while(dvd >= dvs){
            long long tmp = dvs;
            long long cnts = 1;
            while(dvd >= (tmp << 1)){
                tmp <<= 1;
                cnts <<= 1;
            }
            dvd -= tmp;
            res += cnts;
        }
        if(sign * res > INT_MAX || sign * res < INT_MIN) {
            return INT_MAX;
        }
        return sign * res;
    }
};

File: python/leetcode/029.py
class Solution:
    def divide(self, dividend, divisor):
        sign = -1 if (dividend < 0) ^ (divisor < 0) else 1
        dvd, dvs = abs(dividend), abs(divisor)
        INT_MAX, INT_MIN = 2 ** 31 - 1, -2 ** 31
        res = 0
        while dvd >= dvs:
            tmp, cnts = dvs, 1
            while dvd >= (tmp << 1):
                tmp <<= 1
                cnts <<= 1
            dvd -= tmp
            res += cnts
        res *= sign
        return res if INT_MIN <= res <= INT_MAX else INT_MAX